hypothesis-python/tests/conjecture/test_float_encoding.py

```diff
@@ -68,17 +68,21 @@ def test_double_reverse(i):
 @given(st.floats())
 def test_draw_write_round_trip(f):
     d = ConjectureData.for_buffer(bytes(10))
-    flt.write_float(d, f)
+    pp = PrimitiveProvider(d)
+    pp._write_float(f)
+
     d2 = ConjectureData.for_buffer(d.buffer)
-    g = flt.draw_float(d2)
+    pp2 = PrimitiveProvider(d2)
+    g = pp2._draw_float()
```

Comment:
I'm accessing `PrimitiveProvider` directly here in anticipation of the ability to swap `ConjectureData.provider` for a different `Provider` implementation. This may not be necessary depending on how you were envisioning the backend swap occurring.

```python
# option 1: inherit from interface, utils like _draw_float not shared between providers
class Provider(abc.ABC):
    @abc.abstractmethod
    def draw_integer(...): pass

    @abc.abstractmethod
    def draw_float(...): pass

    @abc.abstractmethod
    def draw_boolean(...): pass

    @abc.abstractmethod
    def draw_string(...): pass

    @abc.abstractmethod
    def draw_bytes(...): pass

class PrimitiveProvider(Provider): pass
class CrosshairProvider(Provider): pass


# option 2: inherit from PrimitiveProvider, utils like _draw_float available everywhere
class CrosshairProvider(PrimitiveProvider): pass
```

Reply:
I like this separate style you've used 👍 

Backend swap: we'll probably have `ConjectureData` consult the current `backend` setting (new) and look up the corresponding `PrimitiveProvider` it constructs in from a global `AVAILABLE_BACKENDS: dict[str, str]`.  Then Crosshair et al can [write a plugin](https://hypothesis.readthedocs.io/en/latest/strategies.html#hypothesis-integration-via-setuptools-entry-points) which does `AVAILABLE_BACKENDS["crosshair.hypothesis_backend.CrosshairPrimitiveProvider"] = "user-facing description here"` and we're done!  

There are still some direct uses of `draw_bits()` etc. which don't go through `PrimitiveProvider` around which will break that, but I'd rather ship this PR as soon as possible and then finish the cleanup in a follow-up; I think that's nicer for downstream experimentation.  And of course, shrinking and the database etc. won't work until we finish the later stages of the overhaul anyway.